[deployment] Improved kubernetes deployment

atb00ker opened the pull request:
Changes inspired by deploying on a bare-metal cluster and GKE cluster:
- Add GKE deployment compatibility
- Improved kubernetes documentation
- Simplified documentation
- Added openwisp-nfs server for kubernetes deployments
- Simplified & fixed makefile circular calls
- Other small fixes/improvements

2stacks commented on docs/GCP.md:
```diff
@@ -0,0 +1,38 @@
+    radius.openwisp.org         GLOBAL-IP
+    topology.openwisp.org       GLOBAL-IP
+    freeradius.openwisp.org     REGIONAL-IP-1
+    openvpn.openwisp.org        REGIONAL-IP-2
```
Maybe we should use a place holder or variable like `dashboard.<your.domain>`.

atb00ker replied:
Yes, that sounds much better to avoid any confusion. :smile: 
On it. :+1:

2stacks commented on kubernetes/ExtServices.yml:
```diff
@@ -0,0 +1,61 @@
+  - host: radius.openwisp.org
+    http: *https_rules
+  - host: topology.openwisp.org
+    http: *https_rules
```
Looks like this is in both ExtServices and Ingress.

atb00ker replied:
Yes, when I apply the "Ingress.yml" file on Google Cloud, it doesn't work and keeps complaining.
So, what I had to do is, create "ExtServices.yml" which doesn't work, it only creates an ingress and defines the path, then, I apply the "Certificates.yml" which gets me the letsencrypt certificates for me,
Finally, I apply the "Ingress.yml" which has the same block with
```
  tls:
  - hosts:
        ...
       (hosts)
        ...
    secretName: openwisp-2stacks-crt
``` 
So that I can finally use the site with https! :smile:

2stacks replied:
I see what you mean with the circular references.  It seems like the step should be to;

1. Create the ingress without tls specification.
2. Create the certificate
3. Update the ingress with the tls specification.

That's pretty much what you're doing I guess but wonder if there's a way without duplicating so many lines of yaml.

atb00ker replied:
Indeed, that'll be ideal. Looking into the cert-manager docs for the same. :smile:

atb00ker replied:
Solved! :smile:

atb00ker replied:
Un-solve, the problem is that ingress-gce doesn't let ingress work if it has `tls` and no certificates.
So I had to a patch:
```
kubectl patch ingress/openwisp-http-ingress \
    --patch '{
        "spec": {
            "tls": [
                {
                "hosts": [
                    "dasboard.openwisp.org",
                    "controller.openwisp.org",
                    "radius.openwisp.org",
                    "topology.openwisp.org"
                ],
                "secretName": "openwisp-tls-secret"
                }
            ]
        }
    }'
```

There may be fix for this in future, so creating an issue

2stacks commented on README.md:
```diff
@@ -41,17 +41,21 @@ The configuration for openwisp images using environment variables is available [
+   1.2. On providers you may initiate a cluster as per your desired configurations. (Read the [GCP cluster requirement](docs/GCP.md) to match those requirements in your cluster)
 
-   2.2 `ingress` in `Ingress.yml` should to be your cluster's loadbalancer IPs.
+2. Changes for your cluster:
```
Need to add info regarding updating all domain names.  Domain names currently need to be configured in Ingress.yml, ExtServices.yml, Configmap.yml, Certificates.yml.  Perhaps we should make all other files pull from the Configmap.yml so that domain names only need to be set in one place.

2stacks commented:
Sounds like an order of operations problem.  Feels like we should log this
as an issue.

On Tue, Oct 15, 2019 at 11:11 AM Ajay Tripathi <notifications@github.com>
wrote:

> *@atb00ker* commented on this pull request.
> ------------------------------
>
> In kubernetes/ExtServices.yml
> <https://github.com/openwisp/docker-openwisp/pull/67#discussion_r335013285>
> :
>
> > +apiVersion: extensions/v1beta1
>
> +kind: Ingress
>
> +metadata:
>
> +  name: fanout-ingress
>
> +  annotations:
>
> +    certmanager.k8s.io/cluster-issuer: letsencrypt-prod
>
> +    kubernetes.io/ingress.global-static-ip-name: external-ip-1
>
> +  labels:
>
> +    app: openwisp-nginx
>
> +spec:
>
> +  rules:
>
> +  - host: dashboard.openwisp.org
>
> +    http: &https_rules
>
> +      paths:
>
> +      - backend:
>
> +          serviceName: openwisp-nginx
>
> +          servicePort: 80
>
> +        path: /*
>
> +  - host: controller.openwisp.org
>
> +    http: *https_rules
>
> +  - host: radius.openwisp.org
>
> +    http: *https_rules
>
> +  - host: topology.openwisp.org
>
> +    http: *https_rules
>
>
> Yes, when I apply the "Ingress.yml" file on Google Cloud, it doesn't work
> and keeps complaining.
> So, what I had to do is, create "ExtServices.yml" which doesn't work, it
> only creates an ingress and defines the path, then, I apply the
> "Certificates.yml" which gets me the letsencrypt certificates for me,
> Finally, I apply the "Ingress.yml" which has the same block with
>
>   tls:
>
>   - hosts:
>
>         ...
>
>        (hosts)
>
>         ...
>
>     secretName: openwisp-2stacks-crt
>
>
> So that I can finally use the site with https! 😄
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/openwisp/docker-openwisp/pull/67?email_source=notifications&email_token=ADUNEHFBR2N63GSCKHVEZHLQOXMRNA5CNFSM4JACFQT2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCH77QHA#discussion_r335013285>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADUNEHBXVKT5KZCMHSSBJYDQOXMRNANCNFSM4JACFQTQ>
> .
>

atb00ker commented:
> Sounds like an order of operations problem.  Feels like we should log this
as an issue.

It kinda makes sense that it wouldn't work.
The `secret: openwisp-crt` is created by the "Certificate.yml" upon successful obtaining of certificate and ingress needs that secret file. So, if the secret doesn't exist, I suppose we should expect it to throw errors.

atb00ker commented:
Update: Still working on it, fixed the issue, need to improve documentation last commit: https://github.com/atb00ker/docker-openwisp/commit/1084ff753a3c7cc3ec51dda0e2dac80f3cd8c32b

atb00ker commented:
Changes inspired by deploying on a bare-metal cluster and GKE cluster:
- Add GKE deployment compatibility
- Improved kubernetes documentation
- Simplified documentation
- Added openwisp-nfs server for kubernetes deployments
- Simplified & fixed makefile circular calls
- Other small fixes/improvements

atb00ker commented:
> Sounds like an order of operations problem.  Feels like we should log this
as an issue.

@2stacks I tired a lot of things, but that there is no fix for that atleast right now.

However, I have managed to reduced the repeated code:
Before cert-manager:
```
---
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: openwisp-http-ingress
  annotations:
    kubernetes.io/ingress.global-static-ip-name: openwisp-http-loadbalancer-ip
  labels:
    app: openwisp-nginx
spec:
  backend:
    serviceName: openwisp-nginx
    servicePort: 80
```

After cert-manager gets certificates from this ingress:

```
---
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: openwisp-http-ingress
  annotations:
    kubernetes.io/ingress.global-static-ip-name: openwisp-http-loadbalancer-ip
  labels:
    app: openwisp-nginx
spec:
  backend:
    serviceName: openwisp-nginx
    servicePort: 80
  tls:
  - hosts:
    - dashboard.openwisp.org
    - controller.openwisp.org
    - radius.openwisp.org
    - topology.openwisp.org
    secretName: openwisp-tls-secret
```

atb00ker commented:
Ah, I'll have to change ingress again and add :
```
  rules:
  - http:
      paths:
      - backend:
          serviceName: openwisp-nginx
          servicePort: 80
```
